internal/controller/controllers/agentserviceconfig_controller.go

```diff
@@ -223,108 +223,34 @@ func (r *AgentServiceConfigReconciler) Reconcile(origCtx context.Context, req ct
 		return reconcile.Result{}, nil
 	}
 
-	if instance.DeletionTimestamp.IsZero() {
-		if !controllerutil.ContainsFinalizer(instance, agentServiceConfigFinalizerName) {
-			controllerutil.AddFinalizer(instance, agentServiceConfigFinalizerName)
-			if err := r.Update(ctx, instance); err != nil {
-				log.WithError(err).Error("failed to add finalizer to AgentServiceConfig")
-				return ctrl.Result{Requeue: true}, err
-			}
-		}
-	} else {
-		// do cleanup and remove finalizer
-		statefulSet := &appsv1.StatefulSet{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      imageServiceName,
-				Namespace: r.Namespace,
-			},
-		}
-		if err := r.Get(ctx, client.ObjectKeyFromObject(statefulSet), statefulSet); err != nil && !errors.IsNotFound(err) {
-			log.WithError(err).Error("failed to get image service stateful set for cleanup")
-			return ctrl.Result{Requeue: true}, err
-		}
-		if err := cleanupImageServiceFinalizer(ctx, asc, statefulSet); err != nil {
-			log.WithError(err).Error("failed to cleanup image service stateful set")
-			return ctrl.Result{Requeue: true}, err
-		}
-
-		controllerutil.RemoveFinalizer(instance, agentServiceConfigFinalizerName)
-		if err := r.Update(ctx, instance); err != nil {
-			log.WithError(err).Error("failed to remove finalizer from AgentServiceConfig")
-			return ctrl.Result{Requeue: true}, err
-		}
-
+	// Ensure relevant finalizers exist (cleanup on deletion)
```

Comment:
The order here is important: first you need to check the deletion time and then ensure.